internal/resource_iam_policy.go

```diff
@@ -0,0 +1,204 @@
+//
+// DISCLAIMER
+//
+// Copyright 2022 ArangoDB GmbH, Cologne, Germany
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Copyright holder is ArangoDB GmbH, Cologne, Germany
+//
+
+package provider
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	common "github.com/arangodb-managed/apis/common/v1"
+	iam "github.com/arangodb-managed/apis/iam/v1"
+)
+
+const (
+	// IAM Policy field names
+	iamPolicyURLFieldName         = "url"
+	iamPolicyRoleBindingFieldName = "binding"
+	iamPolicyRoleFieldName        = "role"
+	iamPolicyGroupFieldName       = "group"
+	iamPolicyUserFieldName        = "user"
+)
+
+// resourceIAMPolicy defines an IAM Policy resource.
+func resourceIAMPolicy() *schema.Resource {
+	return &schema.Resource{
+		Description: "Oasis IAM Policy Resource",
+
+		CreateContext: resourceIAMPolicyCreate,
+		ReadContext:   resourceIAMPolicyRead,
+		DeleteContext: resourceIAMPolicyDelete,
+		Schema: map[string]*schema.Schema{
+			iamPolicyURLFieldName: {
+				Type:        schema.TypeString,
+				Description: "IAM Policy Resource IAM Policy URL",
+				ForceNew:    true,
+				Required:    true,
+			},
+			iamPolicyRoleBindingFieldName: {
+				Type:        schema.TypeList,
+				Description: "IAM Policy Resource IAM Policy Bindings",
+				MinItems:    1,
+				ForceNew:    true,
+				Required:    true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						iamPolicyRoleFieldName: {
+							Type:        schema.TypeString,
+							Required:    true,
+							Description: "IAM Policy Resource IAM Policy Role",
+							Elem:        &schema.Schema{Type: schema.TypeString},
+						},
+						iamPolicyGroupFieldName: {
+							Type:        schema.TypeString,
+							Optional:    true,
+							Description: "IAM Policy Resource IAM Policy Group",
+						},
+						iamPolicyUserFieldName: {
+							Type:        schema.TypeString,
+							Optional:    true,
+							Description: "IAM Policy Resource IAM Policy User",
+						},
```

Comment:
How is it enforced that there is always a group OR a user (not both, not neither)?

Reply:
This is done through `ConflictsWith`, but it is a [bug in Terraform](https://github.com/hashicorp/terraform-plugin-sdk/issues/71) and can't be used with TypeLists (MaxItems should be set to 1 and we don't want that in our case). However, I implemented a workaround that makes it impossible to use both/none when creating a role binding (note that it fails during Create execution).